### ISSUE TITLE
Inherited export variables from a registered class don't show up in node inspector

### ISSUE THREAD

**Issue by Lucrecious** [godotengine/godot#23316]
**Godot version:**
Godot 3.1 Alpha

**OS/device including version:**
Mac OSX Sierra 10.12.6

**Issue description:**
Suppose you have a base class in GDScript.

```
# res://base.gd
extends Node2D

class_name Base

export(int) var number : int
```

(1) You can extend by using the path to the .gd script:
```
# res://inherited.gd
extends "res://base.gd"

```

(2) You can extend by using the registered class name:
```
# res://inherited.gd
extends Base
```

in case two the export variable `number` from `Base` does not appear in the inspector.

**Steps to reproduce:**

1.  Create a new Godot project with an empty scene.
2. Add a Node of any kind and give it a script, say "res://base.gd"
3. Give "res://base.gd" an export variable and register its class name
4. Create a new Node that extends from "res://base.gd" but uses the class name instead of the gdscript path
5. The export variable inside "res://base.gd" does not show up in the inherited node inspector

**Minimal reproduction project:**
[class_name_export_bug.zip](https://github.com/godotengine/godot/files/2520965/class_name_export_bug.zip)

**Comment by Nobelix** [comment]
duplicate of #21076 , how ever this one is more detailed

**Comment by DOGEIGHT** [comment]
Exactly same issue I'm having.

**Comment by puzzud** [comment]
I noticed this problem in Godot 3.1.1. Definitely seems a regression, as I think it worked well with pre-"class_name" style extension in Godot 3.0.*.

I had created an extended script (via the **Extend Script** option) on a scene which was already featured in some scene instances in another open scene. This scene also featured instances of the scenes which used the base script. It seemed at runtime, the nodes which had newly received the newly created extension scripts/classes, actually had the exports from the base script (albeit with their default values), but as this Issue states, they did not show up in the inspector of the scene which received the extended script.

I closed out to the Project List and reloaded this project. That seems to have fixed things somewhat, although there is some **data loss**, as the values for the exported variables that had been set prior to the script extension were no longer there.

Sorry if this comment is very verbose. I just hope it can give some insight on this bug.

**Comment by Mastermori** [comment]
This issue is still valid in 3.2beta4.

**Comment by Mastermori** [comment]
And this issues goes even further than not showing the export from the base class, it shows none of the exports from the inherited class.
Assuming the same setup as in the issue:
```gdscript
# res://base.gd
extends Node2D

class_name Base

export(int) var number : int
```
```
# res://inherited.gd
extends Base

export(int) var number2 : int
```
This code shows no script variables when inspecting inherited.tscn, whilst
```
# res://inherited.gd
extends "res://base.gd"

export(int) var number2 : int
```
shows number and number2 as script variables in inspector.

**Comment by Mastermori** [comment]
I don't know when exactly this was fixed, but it is no longer a problem at least scince 3.2beta6. It also works in 3.2rc1 and 3.2rc2, so this issue can be closed

**Comment by akien-mga** [comment]
Thanks, closing.

**Comment by alandoni** [comment]
On v2.3.1.stable.official version, I can reproduce this issue

**Comment by iamiota** [comment]
This problem still exists on Godot 3.2.3.

``` godot
# not show export variables
# Base from autoload
extends Base

# show export variables
extends "res://base.gd"
```

**Comment by berarma** [comment]
Still happening in Godot 3.3.2. Shouldn't this issue be reopened or is it already being tracked elsewhere?

**Comment by scottkunkel** [comment]
confirming it's still broken in 3.4.stable for inherited resource children
restarting the editor is not helping

**Comment by Altoc** [comment]
Here to say this is happening to me in 3.4 stable. 

In my situation, the base class inherits Area, and the child class inherits the base class. None of the export variables are showing in the editor of the child object.

**Comment by dawnseed-dev** [comment]
Still happening with 3.4.2 stable on macOS Monterey 12.2.1. 

Using `extend` with the script path instead of the class name helped for a while, but now it's happening again. 

Even when I manage to get it corrected, the exported variables' values are lost.

**Comment by KoBeWi** [comment]
Can't reproduce in 3.5 beta.

**Comment by KK-MSD** [comment]
3.4.4 stable. This occasionally happens to inheriting node/scene, if it is the parent scene. That is to say, I can always see the export variables properly if the node/scene is instantiated as a child of another scene.

**Comment by Lilith-In-Starlight** [comment]
Issue exists in godot 3.5 stable

**Comment by mishkarch** [comment]
Still present in 3.5.1. After reopening the entire project, the export variables start to appear in the inspector.

**Comment by KoBeWi** [comment]
This is updated project from the OP:
[class_name_export_bug.zip](https://github.com/godotengine/godot/files/11900377/class_name_export_bug.zip)
The issue is not present in 4.1 rc.

@H0nney Please provide a minimal reproduction project.

**Comment by jsonify** [comment]
This is still happening in 4.1.stable. After reloading the project, the export variables appear now, though. I've tried it using both with extending the class_name as well as extending the the "res://...gd" script.

**Comment by SmartASCII** [comment]
Running into the same issue in 4.1.1.stable.

If I extend from the resource path instead of the parent class name it seems to fix it -- even if I then go back to extending from the parent class name afterward.

**Comment by eterlan** [comment]
Hi ! I found that the export var in the base class would appear if I reload scene (Scene -> reload saved scene) OR write any thing in the inherited script. 
Still, auto refresh might be more desirable..

**Comment by good-gamedev** [comment]
Same issue in v.4.1.2.

**Comment by gameoflord012** [comment]
please fix it already

**Comment by YuriSizov** [comment]
@gameoflord012 Godot is developed largely by volunteers, and we have a huge backlog of reports and features to work on as it is. It will be fixed in due time, but please avoid posting entitled comments like that. Consider leaving an upvote on the issue itself to give it more visibility.

**Comment by KoBeWi** [comment]
Reminder that this issue needs a reproduction project. https://github.com/godotengine/godot/issues/23316#issuecomment-1612192496

**Comment by DoodleGarbage** [comment]
I couldn't reproduce this issue with a NEW reproduction project, however:
The original reproduction project attached with the bug report does have this issue present, but editing the script in any way fixes the issue.

Additionally, I believed I had encountered this issue in a different project of my own, but it was just user error by putting a script instead of the resource.

**Comment by YuriSizov** [comment]
@DoodleGarbage Which version did you test it with? Also did you convert the project from the OP yourself or did you use KoBeWi's updated version?

**Comment by DoodleGarbage** [comment]
@YuriSizov I tested using Godot 4.2.1 stable. I tested both projects posted by OP and KoBeWi. The one posted by OP has the bug present, but the one posted by KoBeWi does not, and I could not create a brand new project that had this bug.

**Comment by KoBeWi** [comment]
The OP's script has errors when opened in Godot 4, so obviously the variables will not appear until you fix it.
Then it's safe to assume that this "bug" only occurs when a script is invalid, which is expected. The editor can't determine variables if the script can't be loaded.

**Comment by DoodleGarbage** [comment]
I've spent some time messing around further. Regarding the OP's project: the bug is present after changing the export[int] to @export and saving, any more editing or reloading the engine corrects the issue. 
I also encountered another issue in a different project, where the description for the exported variables were not loaded until the script was edited and saved. I unfortunately could not reproduce this issue.

**Comment by akyoto** [comment]
Just wanted to add that this is not just a faulty script problem, it happened in 4.2.1.stable without an error in the scripts.
Only a full editor reload brought back the inherited export variables.

**Comment by YuriSizov** [comment]
@akyoto If you can reproduce it, please provide a project and exact steps that demonstrate the issue. As stated above, we don't really have much luck reproducing it.

**Comment by brendor** [comment]
I reproduced this but haven't had time to make a minimal repro project. I had both base and inherited script (via class name, not sure if relevant) give me errors, because I was referencing AutoLoad property (compilation failed, there's a separate bug for that). Everything seemed normal after these errors, except this thing where I couldn't see inherited properties (and upon saving scene, the data was lost, I had to manually revert them in the scene file). I had these inherited classes attached to nodes which were added to editable children of a scene instance (of a .glb exported scene). Reimporting the scene/project didn't help. Only after switching `global.<call>` to `get_node("root/global")` to get rid of these compilation errors for both inherited and base classes (both had them) I was able to reimport and see my inherited properties.

**Comment by 1054009** [comment]
On 4.2.2 Mono

`NodeThingA` -> `NodeThingA.cs`
```cs
using Godot;

public partial class NodeThingA : Node3D
{
	[Export]
	public int SomeCoolValue = 100; // Works as expected
}
```

`NodeThingB` ->`NodeThingB.cs`
```cs
using Godot;

public partial class NodeThingB : NodeThingA
{
	[Export]
	public int SomeEvenCoolerValue = 300; // Not shown in inspector
}
```

**Comment by yannick-was-taken** [comment]
While I cannot reproduce the bug exactly as described originally, I can reproduce a variant that seems like the same underlying issue, assumedly.

Please find the reproduction repository at https://github.com/yannick-was-taken/godot-bug-23316 .

Version: v4.2.2.stable.official [15073afe3]

Description:

The scene `Root.tscn` has an attached script `Root.gd` which inherits from named class `Foo` which exports a property `secret`. This property is visible in the inspector for the root node of `Root.tscn`. This **works as expected**.

The scene `RootInner.tscn` has an attached script `RootInner.gd` which inherits from the inner class `Inner` of `Foo`, which exports a property `top_secret`. This propert is **not** visible in the inspector for the root node of `RootInner.tscn`. However, printing the property in `_ready` (not in repo) works, so the property is there. This **deviates from expected behavior**.

**Comment by yannick-was-taken** [comment]
Preliminary investigation results:

When `RootInner.gd` has its exports updated for showing in the inspector, its `base_cache` has an empty `members_cache`. Don't know why yet. On the other hand, `Root.gd`'s `base_cache` has the correct entries in the `members_cache`.
